QuantumRangers-SE-Team-10/lasertech-backend
# Enable player to join a game session and perform all neccessary function

### Summary
The main goal of this pr is to facilitate the ability for a player to join a game session and for the application to perform all the necessary function afterward.

### Details
I start by creating the PlayerSession endpoint, this allow an existing player to join an existing game session. Next I added UDP broadcast as a singleton to both the PlayerSession controller and Game controller in order to start broadcasting a list of equipmentID upon creating a new game session and appending more equipmentID as more PlayerSession is created. Unrelated to this PR, I also added some Docker and Docker-compose file in order to containerize my application and easily deploy it. The benefit of using Docker extends beyond deployment, using Docker they can simply run one command to download and run the latest version of my .NET back-end without having to worry about configuring database or installing the necessary dependency.  

### Design Decision
In order to make the PlayerSession and PlayerSessionDTO class more strongly typed, I used Enums to define a set of named constant to ensure reliability and data integrity. Additionally DTOs are implemented in the controller to simplify and optimize data exchange between front-end and back-end, hiding internal details and complexities which will help with security and scalability. 

### Testing Instruction
1. Ensure that your database is properly connected by sending a Get request to ```/api/game```, it should return an empty array or some data and not an "internal server error" message.

<img width="1455" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/916809b9-5e25-4e16-a583-a7acf0507532">

2. Send a POST request to ```/api/game``` to create a new game session, note the GameID

<img width="1458" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/9078753d-69cf-4364-9282-eb17a21066dc">

3. Onboard a new player by sending a POST request to ```api/player``` or use an existing player, note the PlayerID

<img width="1458" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/9d57a4fb-06fc-43f5-9b0e-03edc76ad4df">

4. Add that player to an existing game session by sending a POST request to ```api/playersession``` sending the correct GameID and the PlayerID in the request body.
<img width="1458" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/dd1313de-828e-4f6e-b167-ad7885b77f19">

5. Ensure that the backend return the correct information along with the latest PlayerSessionID
<img width="1457" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/8cda4563-2cd1-4c46-827d-ec0f4459e3fe">

6. Check the backend console, it should print "Broadcasting Equipment ID..." every second

<img width="1080" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/f0c015c6-c77d-4fe2-8844-c91a89835a61">

7. Run the ```python_udpserver.py``` that Strother provide in his GitHub repository, besure to change the localport to 7500.
<img width="1356" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/1350e1b4-5d10-4aa5-ab83-0e2806fc7e64">

8. Ensure that it repeatedly print the correct Equipment ID
<img width="1004" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/30bb7695-d856-4187-8347-56fd1bc067ea">





### Checks
